### PR TITLE
Unify Graph class hierarchy

### LINKED ISSUE #741: Abstract Graph interface and unify Graph/OptGraph hierarchies
Graph & OptGraph (with GraphNode & OptNode) completely duplicate each other.
The original purpose of separating Optimiser from doman (ML) specific can be acheived by using common Graph interface.

### PR DESCRIPTION
**Pull request by gkirgizov**
That's an important PR that starts big changes in the core of the core: Graph hierarchy design.

My analysis of the current situation with Graphs:
There is a lot of explicit functionality that tries to achieve the following points:
1. Draw a clear boundary between Optimiser & problem domain (AutoML/EPDE in future/etc) -- done through duplicated Graph & OptGraph classes (same for GraphNode & OptNode classes)
2. Share single Graph implementation -- done through GraphOperator & NodeOperator classes. Graph/OptGraph just delegated these calls to Operators.

Yet this approach is not without its problems:
- Graph & OptGraph still duplicate code, that complicates maintenance and further development of the framework core.
- GraphOperator requires cyclic reference to the owned graph, that complicates some code (e.g. serialization).
- Adapter functionality becomes more complex without good reason -- in effect, it just implements runtime transformations between classes where they could be avoided altogether (i.e. Pipeline is already a Graph, which is the same as OptGraph)

Same points can be achieved with a cleaner design: What previously was done manually (sharing of graph implementation between OptGraph & Graph through the GraphOperator) could be actually encoded in class hierarchy. So now instead of runtime transformations and delegation to GraphOperator everything is encoded in class hierarchy, while implementation details are hidden behind opaque Graph interface.

----

Additional benefits:
- We can have more (and maybe external) Graph implementations. For example, now it's in principle possible to add **network-x** Graph implementation, which would give us more guarantees in correctness and efficiency of graph operations.

The continuation of this work will be #742 . *Place of Adapters will be preserved* and we still will be able to optimise completely foreign structures -- as far as they can be translated to Graph. But now framework users will have additional choice: *if their optimised structure conforms to the Graph interface, no adaptation is required, as Optimiser will work through the Graph interface.*

Specific notes on my proposed changes:
- Graph interface is introduced
- GraphOperator becomes the only implementation of Graph
- GraphDelegate provides inheritance through composition -- in effect, substitutes what was done previously through delegation to `self.operator`
- OptGraph now is just a wrapper that adds `self.log` to Graph -- could remove it after #622, and just make it an alias of GraphDelegate
- Introduce `Copyable` mixin-class. GraphDelegate & GraphOperator inherit from it. (that's instead of direct impl of copy & deepcopy in OptGraph)
- OptNode is just an alias for GraphNode
- NodeOperator is dropped (now it's unnecessary because there's only one GraphNode impl)
- Avoid the need for explicit passing of `DEFAULT_PARAMS_STUB` in OptNode and slightly refactor handling of it

Fixes #741

### CODE REVIEW

File: fedot/core/dag/graph_operator.py [hunk 1]
```diff
@@ -1,39 +1,51 @@
...
 
 
-class GraphOperator:
-    def __init__(self, graph=None, postproc_nodes=None):
-        self._graph = graph
-        self._postproc_nodes = postproc_nodes
+NodePostprocessCallable = Callable[[Graph, Sequence[GraphNode]], Any]
+
+
+class GraphOperator(Graph, Copyable):
```
Comment (nicl-nno) [inline fedot/core/dag/graph_operator.py:16]: Тут стоит надо пояснить, зачем GraphOperator нужен.
И его название точно актуально? Просто ниже у тебя есть 

`GraphImpl = GraphOperator`

что усложняет понимание разницы между GraphOperator/Graph.

Мб их надо назвать Graph/AbstactGraph?

File: fedot/core/dag/graph.py [hunk 2]
```diff
@@ -1,97 +1,115 @@
...
-        self.operator.delete_subtree(subroot)
-
-    def show(self, path: str = None):
-        GraphVisualiser().visualise(self, path)
+        raise NotImplementedError()
 
+    @abstractmethod
     def __eq__(self, other) -> bool:
-        return self.operator.is_graph_equal(other)
+        # return self.operator.is_graph_equal(other)
```
Comment (maypink) [inline fedot/core/dag/graph.py:66]: вероятно забывашка

File: fedot/core/dag/graph_operator.py (hunk 1 above)
Comment (gkirgizov) [reply fedot/core/dag/graph_operator.py:16]: Да, переименую перед мержем. Или маленьким отдельным PR — чтобы этот не засорять кучей мелких изменений.
GraphOperator не очень актуален, но опять же — это чтобы пока поменьше несущественных изменений видеть.

**Review by nicl-nno — APPROVED** [review]

File: fedot/core/dag/graph_delegate.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    """
+    Graph that delegates calls to another Graph implementation.
+
+    The class purpose is for cleaner code organisation:
+    - avoid inheriting from specific Graph implementations
+    - hide Graph implementation details from inheritors.
+    """
+
+    def __init__(self, *args, delegate_cls: Type[Graph] = GraphOperator, **kwargs):
+        self.operator = delegate_cls(*args, **kwargs)
```
Comment (MorrisNein) [inline fedot/core/dag/graph_delegate.py:19]: Такой вопрос: а почему бы не дать пользователю инициализировать инстанс извне? Более гибкий вариант использования получается. Типа метода from_instance или вообще перегрузкой инита.

Comment (MorrisNein) [reply fedot/core/dag/graph_delegate.py:19]: И ещё один вопрос. На твой взгляд, упростит ли нам жизнь прокинуть обращение к любым атрибутам объекта `GraphDelegate` на внутренний инстанс другого класса?

Comment (MorrisNein) [reply fedot/core/dag/graph_delegate.py:19]: Мне кажется, что имя `operator` плохо подходит для такого использования. Ранее это был объект отдельного класса с единственным назначением, а теперь это некий объект с вообще любой логикой.

Если это не временная мера, чтобы избежать переименования `operator` по всему проекту, предлагаю дать атрибуту новое имя. Я предлагаю `graph_implementation`.

**Pull request merged**